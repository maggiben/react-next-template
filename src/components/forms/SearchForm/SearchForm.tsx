import { useState, useRef, ReactNode } from 'react';
import {
  Button,
  Grid,
  GridItem,
  Link,
  TextInput,
  Select,
  SearchIcon,
  MailIcon,
} from "@fravega-it/bumeran-ds-fvg";
import { useTranslation } from 'react-i18next';
import string from '@utils/string';
import * as yup from 'yup';
import styled from "styled-components";

const FormContainer = styled.div`
  border-radius: ${({ theme }) => theme.borderRadius.m};
  width: 100%;
  background-color: ${({ theme }) => theme.colors.neutral[100]};
  display: flex;
  padding: 16px;
  flex-direction: column;
  align-items: flex-start;
  gap: 16px;
`;

const SpaceTop = styled.div<{ size: 'xxxs' | 'xxs' | 'xs' | 's' | 'm' | 'l' | 'xl' | 'xxl' | 'xxxl'; }>`
  margin-top: ${({ theme, size }) => theme.spacing[size]};
`;

export type FormValues = {
  cuid?: string;
  documentType?: string;
  documentNumber?: string;
  email?: string;
}
  
type SearchFormProps = {
  onSearch: (data: FormValues) => void;
  documentType?: string;
  search?: string;
  cuid?: number;
  email?: string;
}

const SearchForm = (props: SearchFormProps): JSX.Element => {
  const { t } = useTranslation();
  const [search, setSearch] = useState<string>('');
  const [documentType, setDocumentType] = useState<string>('dni');
  const [searchLabel, setSearchLabel] = useState<string>(t('dni'));
  const [errorMessage, setErrorMessage] = useState<string>('');
  
  const { onSearch } = props;

  const handleKeyDown = (key: string) => {
    // Check if the pressed key is Enter (key code 13)
    if (key === 'Enter') {
      handleOnSearch();
    }
  };

  const handleReset = (event: React.MouseEvent<HTMLAnchorElement>) => {
    setSearch('');
  }

  const handleChange = (value: string) => {
    setSearch(value);
  }

  const handleDocumentTypeChange = ({id, label}: {id: string; label: string}) => {
    setDocumentType(id);
    setSearchLabel(label);
  };

  const handleOnSearch = () => {
    switch(documentType) {
      case 'dni': {
        onSearch({
          documentType,
          documentNumber: search,
        });
        break;
      }
      case 'email': {
        const isEmail = string.validateEmail(search);
        if(!isEmail) {
          setErrorMessage(t('invalid email'));
        } else {
          setErrorMessage('')
          onSearch({
            email: search,
          });
        }
        break;
      }
    }
  }
  
  return (
    <FormContainer>
      <div style={{ width: '100%'}}>
        <Grid>
          <GridItem xs={4}>
            <div data-testid="select-document-type">
              <Select
                data-testid="documentType"
                name='documentType'
                id="documentType"
                label={`${t('search by')}...`}
                options={[
                  { id: 'dni', label: t('dni') },
                  { id: 'email', label: t('email') },
                ]}
                onChange={handleDocumentTypeChange}
                value={documentType}
              />
            </div>
          </GridItem>
          <GridItem xs={4}>
            <TextInput
              leftIcon={documentType === 'dni' ? <SearchIcon /> : <MailIcon />}
              id="search"
              name="search"
              label={searchLabel} 
              onChange={handleChange}
              onKeyDown={handleKeyDown}
              value={search}
              error={Boolean(errorMessage.length)}
              errorMessage={errorMessage}
            />
          </GridItem>
          
          <GridItem xs={4} alignSelf="end" justifySelf="start">
            <Button
              label={t('search')} 
              variant="primary"
              onClick={handleOnSearch}
            />
          </GridItem>
        </Grid>
        <SpaceTop size="s" />
        <Grid>
          <GridItem xs={4} alignSelf="end" justifySelf="start">
            <Link>
              <a onClick={handleReset}>{t('clan filters')}</a>
            </Link>
          </GridItem>
        </Grid>
      </div>
    </FormContainer>
  );
};

export default SearchForm;