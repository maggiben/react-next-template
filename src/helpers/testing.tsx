import { default as baseRenderer } from 'react-test-renderer';
import { render, RenderResult } from '@testing-library/react';
import i18n from '@utils/i18n';
import { I18nextProvider } from 'react-i18next';
import { RecoilRoot } from 'recoil';
import { ThemeProvider } from 'styled-components';
import { theme, IDefaultTheme } from '@fravega-it/bumeran-ds-fvg';
import React, { ReactElement } from 'react';
import '@testing-library/jest-dom'

/**
 * Helper used to render components for testing.
 * The function appends the theme to the children.
 * */
interface IRenderer {
  create: (children: ReactElement) => RenderResult;
  baseRenderer: (children: ReactElement) => baseRenderer.ReactTestRenderer;
  renderWithTheme: (children: ReactElement, theme: IDefaultTheme) => RenderResult;
  renderWithI18n: (children: ReactElement) => RenderResult;
}

export const renderer: IRenderer = {
  create: (children: ReactElement): RenderResult =>
    render(
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
  ),
  renderWithTheme: (children: ReactElement, theme: IDefaultTheme): RenderResult =>
    render(
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
  ),
  baseRenderer: (children: ReactElement): baseRenderer.ReactTestRenderer =>
    baseRenderer.create(
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
  ),
  renderWithI18n: (children: ReactElement): RenderResult => 
    render(
      <ThemeProvider theme={theme}>
        <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
      </ThemeProvider>
  ),
};