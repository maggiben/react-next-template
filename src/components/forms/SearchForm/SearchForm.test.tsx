import React from 'react';
import { renderer } from '@helpers/testing';
import { fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import SearchForm from './SearchForm';
import { expect } from '@jest/globals';
import i18next from 'i18next';

describe('SearchForm', () => {
  it('renders the SearchForm component', () => {
    const onSearch = jest.fn();
    const tree = renderer.renderWithI18n(<SearchForm onSearch={onSearch} />);
    expect(tree).toMatchSnapshot();
  });

  it('search with input', () => {
    const onSearch = jest.fn();
    const tree = renderer.renderWithI18n(<SearchForm onSearch={onSearch} />);
    const searchInput = tree.getByRole('textbox', { name: i18next.t('dni') });
    const searchButton = tree.getByRole('button', { name: i18next.t('search') });//, { name: 'Accept' })[0];

    fireEvent.click(searchButton);
    expect(onSearch).toHaveBeenCalledWith({
      documentNumber: '',
      documentType: 'dni',
    });
  });
  
  it('search with input with value', () => {
    const onSearch = jest.fn();
    const dni = '3000001';
    const tree = renderer.renderWithI18n(<SearchForm onSearch={onSearch} />);
    const searchInput = tree.getByRole('textbox', { name: i18next.t('dni') });
    const searchButton = tree.getByRole('button', { name: i18next.t('search') });//, { name: 'Accept' })[0];

    fireEvent.change(searchInput, { target: { value: dni } });
    fireEvent.click(searchButton);
    expect(onSearch).toHaveBeenCalledWith({
      documentNumber: dni,
      documentType: 'dni',
    });
  });

  it('search with input with value', () => {
    const onSearch = jest.fn();
    const documentNumenr = '3000001';
    const email = 'tom@fravega.com.ar';
    const documentType = 'email';
    const tree = renderer.renderWithI18n(<SearchForm onSearch={onSearch} />);
    
    const select = tree.getByText(`${i18next.t('search by')}...`);
    const searchInput = tree.getByRole('textbox', { name: i18next.t('dni') });
    const searchButton = tree.getByRole('button', { name: i18next.t('search') });//, { name: 'Accept' })[0];
    const container = tree.getByTestId('select-document-type');
    const input = container.querySelector('input');
    
    fireEvent.change(searchInput, { target: { value: email } });

    if (input) {
      fireEvent.click(input);
      const menu = container.firstChild?.firstChild?.childNodes[1];
      const dniItem = menu?.firstChild?.childNodes[0];
      const emailItem = menu?.firstChild?.childNodes[1];
      if (emailItem) {
        fireEvent.click(emailItem);
        fireEvent.click(searchButton);
        expect(onSearch).toHaveBeenCalledWith({
          email,
        });
      }
    }
  });
});
