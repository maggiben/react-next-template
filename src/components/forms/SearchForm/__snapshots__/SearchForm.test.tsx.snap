// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm renders the SearchForm component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="SearchForm__FormContainer-sc-1w9hh7q-0 ceYvVq"
      >
        <div
          style="width: 100%;"
        >
          <div
            class="sc-jsJBEP gDratY"
          >
            <div
              class="sc-eeDRCY eAaUXa"
            >
              <div
                style="display: contents;"
              >
                <div
                  class="sc-bbSZdi bYQRsd"
                >
                  <div
                    class="sc-eBMEME hbpwPF"
                  >
                    <label
                      class="sc-dtInlm bmExAe"
                      for="documentType_input"
                    >
                      <div
                        aria-level="5"
                        class="sc-kAyceB fTWNaB"
                        color="inherit"
                        role="heading"
                      >
                        Buscar por...
                      </div>
                    </label>
                    <div
                      class="sc-cWSHoV sc-dCFHLb kchOeC dZdvSt"
                    >
                      <div
                        style="width: 100%;"
                      >
                        <span
                          class="sc-iGgWBj cifpBe"
                        >
                          <input
                            autocomplete="off"
                            class="sc-fhzFiK cLeJjZ"
                            id="documentType_input"
                            readonly=""
                            type="text"
                            value="DNI"
                          />
                        </span>
                      </div>
                      <div
                        class="sc-jxOSlx clLfZS"
                      >
                        <button
                          class="sc-lcIPJg fQTKST"
                        >
                          <svg
                            class="sc-kpDqfm gtIuVz"
                            color="black"
                            fill="none"
                            viewBox="0 0 28 28"
                            width="24px"
                          >
                            <path
                              clip-rule="evenodd"
                              d="M8.006 11.228a.778.778 0 0 1 1.1 0L14 16.122l4.895-4.894a.778.778 0 0 1 1.1 1.1l-5.445 5.444a.778.778 0 0 1-1.1 0l-5.444-5.444a.778.778 0 0 1 0-1.1Z"
                              fill-rule="evenodd"
                            />
                          </svg>
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="sc-eeDRCY eAaUXa"
            >
              <div
                class="sc-eBMEME hbpwPF"
              >
                <label
                  class="sc-dtInlm bmExAe"
                  for="search"
                >
                  <div
                    aria-level="5"
                    class="sc-kAyceB fTWNaB"
                    color="inherit"
                    role="heading"
                  >
                    DNI
                  </div>
                </label>
                <div
                  class="sc-cWSHoV sc-dCFHLb kchOeC dZdvSt"
                >
                  <div
                    class="sc-jxOSlx clLfZS"
                  >
                    <button
                      class="sc-lcIPJg fQTKST"
                    >
                      <svg
                        class="sc-kpDqfm gtIuVz"
                        color="black"
                        fill="none"
                        viewBox="0 0 28 28"
                        width="24px"
                      >
                        <path
                          clip-rule="evenodd"
                          d="M12.444 7.778a4.667 4.667 0 1 0 0 9.333 4.667 4.667 0 0 0 0-9.333Zm-6.222 4.667A6.222 6.222 0 1 1 17.36 16.26l4.19 4.19a.778.778 0 1 1-1.1 1.1l-4.19-4.19a6.222 6.222 0 0 1-10.038-4.915Z"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </button>
                  </div>
                  <div
                    style="width: 100%;"
                  >
                    <span
                      class="sc-iGgWBj cifpBe"
                    >
                      <input
                        class="sc-fhzFiK cLeJjZ"
                        id="search"
                        type="text"
                        value=""
                      />
                    </span>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="sc-eeDRCY cYwUmi"
            >
              <button
                class="sc-fHjqPf cEiVYx"
              >
                <label
                  aria-label="Buscar"
                >
                  <div
                    aria-level="5"
                    class="sc-kAyceB fTWNaB"
                    color="inherit"
                    role="heading"
                  >
                    Buscar
                  </div>
                </label>
              </button>
            </div>
          </div>
          <div
            class="SearchForm__SpaceTop-sc-1w9hh7q-1 hxISia"
          />
          <div
            class="sc-jsJBEP gDratY"
          >
            <div
              class="sc-eeDRCY cYwUmi"
            >
              <span
                class="sc-koXPp dsIkix"
                tabindex="0"
              >
                <span
                  class="sc-iGgWBj dBDQUE"
                  color="inherit"
                >
                  <a
                    tabindex="-1"
                  >
                    Limpiar filtros
                  </a>
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="SearchForm__FormContainer-sc-1w9hh7q-0 ceYvVq"
    >
      <div
        style="width: 100%;"
      >
        <div
          class="sc-jsJBEP gDratY"
        >
          <div
            class="sc-eeDRCY eAaUXa"
          >
            <div
              style="display: contents;"
            >
              <div
                class="sc-bbSZdi bYQRsd"
              >
                <div
                  class="sc-eBMEME hbpwPF"
                >
                  <label
                    class="sc-dtInlm bmExAe"
                    for="documentType_input"
                  >
                    <div
                      aria-level="5"
                      class="sc-kAyceB fTWNaB"
                      color="inherit"
                      role="heading"
                    >
                      Buscar por...
                    </div>
                  </label>
                  <div
                    class="sc-cWSHoV sc-dCFHLb kchOeC dZdvSt"
                  >
                    <div
                      style="width: 100%;"
                    >
                      <span
                        class="sc-iGgWBj cifpBe"
                      >
                        <input
                          autocomplete="off"
                          class="sc-fhzFiK cLeJjZ"
                          id="documentType_input"
                          readonly=""
                          type="text"
                          value="DNI"
                        />
                      </span>
                    </div>
                    <div
                      class="sc-jxOSlx clLfZS"
                    >
                      <button
                        class="sc-lcIPJg fQTKST"
                      >
                        <svg
                          class="sc-kpDqfm gtIuVz"
                          color="black"
                          fill="none"
                          viewBox="0 0 28 28"
                          width="24px"
                        >
                          <path
                            clip-rule="evenodd"
                            d="M8.006 11.228a.778.778 0 0 1 1.1 0L14 16.122l4.895-4.894a.778.778 0 0 1 1.1 1.1l-5.445 5.444a.778.778 0 0 1-1.1 0l-5.444-5.444a.778.778 0 0 1 0-1.1Z"
                            fill-rule="evenodd"
                          />
                        </svg>
                      </button>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="sc-eeDRCY eAaUXa"
          >
            <div
              class="sc-eBMEME hbpwPF"
            >
              <label
                class="sc-dtInlm bmExAe"
                for="search"
              >
                <div
                  aria-level="5"
                  class="sc-kAyceB fTWNaB"
                  color="inherit"
                  role="heading"
                >
                  DNI
                </div>
              </label>
              <div
                class="sc-cWSHoV sc-dCFHLb kchOeC dZdvSt"
              >
                <div
                  class="sc-jxOSlx clLfZS"
                >
                  <button
                    class="sc-lcIPJg fQTKST"
                  >
                    <svg
                      class="sc-kpDqfm gtIuVz"
                      color="black"
                      fill="none"
                      viewBox="0 0 28 28"
                      width="24px"
                    >
                      <path
                        clip-rule="evenodd"
                        d="M12.444 7.778a4.667 4.667 0 1 0 0 9.333 4.667 4.667 0 0 0 0-9.333Zm-6.222 4.667A6.222 6.222 0 1 1 17.36 16.26l4.19 4.19a.778.778 0 1 1-1.1 1.1l-4.19-4.19a6.222 6.222 0 0 1-10.038-4.915Z"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </button>
                </div>
                <div
                  style="width: 100%;"
                >
                  <span
                    class="sc-iGgWBj cifpBe"
                  >
                    <input
                      class="sc-fhzFiK cLeJjZ"
                      id="search"
                      type="text"
                      value=""
                    />
                  </span>
                </div>
              </div>
            </div>
          </div>
          <div
            class="sc-eeDRCY cYwUmi"
          >
            <button
              class="sc-fHjqPf cEiVYx"
            >
              <label
                aria-label="Buscar"
              >
                <div
                  aria-level="5"
                  class="sc-kAyceB fTWNaB"
                  color="inherit"
                  role="heading"
                >
                  Buscar
                </div>
              </label>
            </button>
          </div>
        </div>
        <div
          class="SearchForm__SpaceTop-sc-1w9hh7q-1 hxISia"
        />
        <div
          class="sc-jsJBEP gDratY"
        >
          <div
            class="sc-eeDRCY cYwUmi"
          >
            <span
              class="sc-koXPp dsIkix"
              tabindex="0"
            >
              <span
                class="sc-iGgWBj dBDQUE"
                color="inherit"
              >
                <a
                  tabindex="-1"
                >
                  Limpiar filtros
                </a>
              </span>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
